Stop the timer when the game is not visible

LeonardoBraga opened the pull request:
- Use the Page Visibility API to abort the timer increment when `document.visibilityState` is `hidden`
- Tweaked the time span logic to allow for pause/resume

Closes #477

jakearchibald commented on src/services/preact-canvas/components/top-bar/index.tsx:
```diff
@@ -65,11 +65,17 @@ class Time extends Component<TimeProps, {}> {
+
       requestAnimationFrame(() => {
-        this.base!.textContent = minSec(Date.now() - this._start!);
+        this.base!.textContent = minSec(++this._secondsSinceStart! * 1000);
```
This assumes that setInterval is accurate, but it isn't.

The correct way to do this (afaik):

* `_start` becomes `_anchorTime`.
* On `_startTimer`, set `_anchorTime` to `Date.now()` & start the interval.
* When visibility becomes hidden, clear the interval, and set `_hiddenTime` to `Date.now()`.
* When visibility becomes visible, increase `_anchorTime` by `Date.now() - _hiddenTime`, and restart the interval.

LeonardoBraga replied:
Thanks for the review, @jakearchibald. In my initial tests, I noticed that Chrome does account for the discrepancies in the `setInterval` by offsetting the next "tick" by enough milliseconds to compensate for the previous ones, but indeed the users of other browsers would get ~1s of unfair advantage every ~16m.

I applied the fixes as you requested.

jakearchibald commented on src/services/preact-canvas/components/top-bar/index.tsx:
```diff
@@ -58,25 +62,45 @@ class Time extends Component<TimeProps, {}> {
+    this.base!.textContent = minSec(Date.now() - this._anchorTime!);
+  }
+
+  private _visibilityChanged() {
```
We have a `@bind` decorator for this, but I'll add this.

LeonardoBraga replied:
Oh. I forgot about that. If you don't mind, let me make the change as part of my next push so I can practice this trick too.

LeonardoBraga replied:
I've pushed the change to use `@bind`, and I'll be waiting for your replies on my follow-ups below.

jakearchibald commented on src/services/preact-canvas/components/top-bar/index.tsx:
```diff
@@ -58,25 +62,45 @@ class Time extends Component<TimeProps, {}> {
+  private _startTimer(resetAnchorTime = true) {
+    if (resetAnchorTime) {
+      this._anchorTime = Date.now();
+    }
```
Let's avoid the boolean trap here by moving `this._anchorTime = Date.now();` to where `startTimer` is called in `componentDidMount`.

LeonardoBraga replied:
This will result in having to set `_anchorTime` twice, as `_startTimer` is called in two places. The boolean trap felt a cleaner solution because of that. If you don't like the implicit default value, we can remove the default and make it explicit.

jakearchibald replied:
> This will result in having to set `_anchorTime` twice, as `_startTimer` is called in two places.

Ah, I didn't see that. Instead, create a `_resetTimer` method, which would be called in places you'd currently call `_startTimer()`/`_startTimer(true)`.

LeonardoBraga replied:
Indeed your suggestion resulted in a cleaner implementation. I left the name `_startTimer` for the method that resets the `_anchorTime` and added a `_resumeTimer` for the one that just resets the interval, as it felt more semantically adequate (`start` continuing as the counterpart of `stop` and `resume` handling an eventual continuation), but I can rename them on a last push if you didn't like the result.

jakearchibald commented on src/services/preact-canvas/components/top-bar/index.tsx:
```diff
@@ -58,25 +62,45 @@ class Time extends Component<TimeProps, {}> {
+      this._stopTimer();
+    } else {
+      this._anchorTime! += Date.now() - this._hiddenTime!;
+      this._updateText();
```
This shouldn't result in a change, right?

LeonardoBraga replied:
It actually does. If you leave the game when you're on the "edge" of another tick and return to it, the clock sometimes jumps two seconds, because the previously stored "almost 1s" and the extra second it takes for the interval block to kick-in results in a 2s difference.

jakearchibald replied:
Fair enough!

LeonardoBraga commented:
Hey @jakearchibald, are you kidding? :-) We should definitely do it in the best way we possibly can!

I had noticed the main timer when I was analyzing the code but I just wasn't sure if I should have gone for a PR what would touch multiple parts of the app, this is why I went for the low hanging fruit. I'm glad you're open to a more thorough change and will take a deeper look at your suggestion, the code, and work on the next iteration.

jakearchibald commented:
We've changed the build for this project which has involved moving a lot of files around. Don't worry about the conflicts, we'll take care of it when we come to merge this.

LeonardoBraga commented:
Sure, thanks for the update. I apologize for the radio silence over the last days, things ended up piling up on my end. I plan on wrapping the modifications you requested this week and send for your review.

LeonardoBraga commented:
@jakearchibald I tried to incorporate the points you raised before, and besides any other feedback you may have, I wanted to call your attention to the following items:
- I've re-used `completeTime` to display the partial timer and removed `endTime`, which wasn't being used. Let me know if you'd rather a different state field.
- I've added a switch case "fall through" to save some bytes in the final bundle by avoiding another call to `stopTimer` and a break. I know it's an unpopular approach, to say the least, so just say the word and I'll remove it.

LeonardoBraga commented:
Hey @jakearchibald, just wanted to know if you had time to check out the latest changes and have any additional feedback. Thanks! :)

jakearchibald commented:
Thanks for continuing with this, and sorry for the slow response.

This is now doing the DOM-diffing every second, which we were trying to avoid due to performance issues on low-end devices. I'll try and think of a way around it.
